### PR TITLE
kubelet/cadvisor: Refactor cadvisor disk stat/usage interfaces.

### PR DESCRIPTION
**Pull request by sjpotter**
basically

1) cadvisor struct will know what runtime the kubelet is, passed in via additional argument to New()

2) rename cadvisor wrapper function to DockerImagesFsInfo() to ImagesFsInfo() and have linux implementation choose a label based on the runtime inside the cadvisor struct

2a) mock/fake/unsupported modified to take the same additional argument in New()

3) kubelet's wrapper for the cadvisor wrapper is renamed in parallel

4) make all tests use new interface

### CODE REVIEW

**Comment by sjpotter** [comment]
@yifan-gu (also who else?)

**Comment by yifan-gu** [comment]
cc @kubernetes/sig-node

**Comment by yifan-gu** [comment]
@sjpotter The commit message needs to be fixed.

File: pkg/kubelet/container/runtime.go [hunk 1]
```diff
@@ -97,6 +97,8 @@ type Runtime interface {
 	RemoveImage(image ImageSpec) error
 	// Returns Image statistics.
 	ImageStats() (*ImageStats, error)
+	// Cadvisor FS Label
+	ImageLabel() string
```
Comment (yifan-gu) [inline pkg/kubelet/container/runtime.go]: nit, period at the end.

File: pkg/kubelet/container/runtime.go [hunk 2]
```diff
@@ -97,6 +97,8 @@ type Runtime interface {
 	RemoveImage(image ImageSpec) error
 	// Returns Image statistics.
 	ImageStats() (*ImageStats, error)
+	// Cadvisor FS Label
```
Comment (euank) [inline pkg/kubelet/container/runtime.go]: godoc style (beginning with `ImageLabel`) and a little more information would be nice.

File: pkg/kubelet/container/runtime.go (hunk 1 above)
Comment (yifan-gu) [inline pkg/kubelet/container/runtime.go]: @sjpotter Actually, does this has to be implemented by runtime? Can it be implemented by kubelet?
e.g. when kubelet starts, it should be able to tell what runtime to use, so then it can save the image label in a variable.

**Comment by yifan-gu** [comment]
Reviewed.

File: pkg/kubelet/container/runtime.go (hunk 1 above)
Comment (euank) [inline pkg/kubelet/container/runtime.go]: I'm not sure I like the use of the term `Label` here. I noticed that the possible values are `root | docker-images | rkt-images`. I think `FSType` or similar might be clearer, here and everywhere else. Second opinions welcome since this is a completely subjective thing.

Comment (sjpotter) [inline pkg/kubelet/container/runtime.go]: it should still have to be provided by the runtime as its ot just the runtime's name, it's cadvisor's label

**Comment by euank** [comment]
small nits, +1 on @yifan-gu comment that the commit message should be improved.

File: pkg/kubelet/container/runtime.go (hunk 1 above)
Comment (sjpotter) [inline pkg/kubelet/container/runtime.go]: I think it makes sense to be part of the runtime, as its not a simple string, but a cadvisor provided label

Comment (yifan-gu) [inline pkg/kubelet/container/runtime.go]: @sjpotter I know it's not runtime's name, I mean kubelet should be able to decide the label by itself when given the container runtime's name.
It just seems unnecessary to have runtime implements this.

**Comment by sjpotter** [comment]
@k8s-bot test this

**Comment by yifan-gu** [comment]
`pkg/kubelet/rkt/rkt.go:1888: undefined: fs.LabelRktImages`

Guess you need a cadvisor release :)

**Comment by vishh** [comment]
Is this needed for v1.3?

**Comment by yifan-gu** [comment]
@vishh Yes, added rktnetes-v1.0 label.

File: pkg/kubelet/cadvisor/cadvisor_linux.go [hunk 3]
```diff
@@ -163,11 +163,22 @@ func (cc *cadvisorClient) MachineInfo() (*cadvisorapi.MachineInfo, error) {
...
+		label = cadvisorfs.LabelRktImages
+	default:
+		return cadvisorapiv2.FsInfo{}, fmt.Errorf("ImagesFsInfo: unknown runtime: %v", runtime)
+	}
+
+	return cc.getFsInfo(label)
 }
 
-func (cc *cadvisorClient) RootFsInfo() (cadvisorapiv2.FsInfo, error) {
+func (cc *cadvisorClient) RootFsInfo(_ string) (cadvisorapiv2.FsInfo, error) {
```
Comment (timstclair) [inline pkg/kubelet/cadvisor/cadvisor_linux.go]: Why did you add a string parameter that's never used?

Comment (sjpotter) [inline pkg/kubelet/cadvisor/cadvisor_linux.go]: We used to have DockerImagesFsInfo() and RootFsInfo()

both were used as arguments to a function.  DockerImagesFsInfo was renamed to ImagesFsInfo(runtime string) so in order to not duplicate code, modified RootFsInfo() to keep the same prototype as it

Comment (sjpotter) [inline pkg/kubelet/cadvisor/cadvisor_linux.go]: I'm not particurally happy with it, but its needed as my previous approach of having a single function with a cadvisorfs label passed in (which would include root) doesn't work for unsupported archs (i.e. darwin, windows....)

**Comment by sjpotter** [comment]
@yifan-gu

File: pkg/kubelet/disk_manager.go [hunk 4]
```diff
@@ -83,15 +83,19 @@ func (dm *realDiskSpaceManager) getFsInfo(fsType string, f func() (cadvisorapi.F
...
+	return dm.isSpaceAvailable(runtime, dm.policy.DockerFreeDiskMB, dm.cadvisor.ImagesFsInfo)
 }
 
 func (dm *realDiskSpaceManager) IsRootDiskSpaceAvailable() (bool, error) {
-	return dm.isSpaceAvailable("root", dm.policy.RootFreeDiskMB, dm.cadvisor.RootFsInfo)
+	return dm.isSpaceAvailable("root", dm.policy.RootFreeDiskMB, dm.rootFsWrapper)
 }
 
-func (dm *realDiskSpaceManager) isSpaceAvailable(fsType string, threshold int, f func() (cadvisorapi.FsInfo, error)) (bool, error) {
+func (dm *realDiskSpaceManager) rootFsWrapper(_ string) (cadvisorapi.FsInfo, error) {
```
Comment (yifan-gu) [inline pkg/kubelet/disk_manager.go]: Another way of getting rid of this, and also keeping the old function signature of `isSpaceAvailable()` is the add the runtime type in  
```go
type cadvisorClient struct {
        runtimeType string
 	manager.Manager
 }
```

File: pkg/kubelet/cadvisor/cadvisor_linux.go [hunk 5]
```diff
@@ -163,8 +163,19 @@ func (cc *cadvisorClient) MachineInfo() (*cadvisorapi.MachineInfo, error) {
 	return cc.GetMachineInfo()
 }
 
-func (cc *cadvisorClient) DockerImagesFsInfo() (cadvisorapiv2.FsInfo, error) {
-	return cc.getFsInfo(cadvisorfs.LabelDockerImages)
+func (cc *cadvisorClient) ImagesFsInfo(runtime string) (cadvisorapiv2.FsInfo, error) {
```
Comment (vishh) [inline pkg/kubelet/cadvisor/cadvisor_linux.go]: We should push this to runtimes ideally. This PR is fine.

File: pkg/kubelet/disk_manager.go [hunk 6]
```diff
@@ -83,15 +83,19 @@ func (dm *realDiskSpaceManager) getFsInfo(fsType string, f func() (cadvisorapi.F
 	return fsi, nil
 }
 
-func (dm *realDiskSpaceManager) IsDockerDiskSpaceAvailable() (bool, error) {
-	return dm.isSpaceAvailable("docker", dm.policy.DockerFreeDiskMB, dm.cadvisor.DockerImagesFsInfo)
+func (dm *realDiskSpaceManager) IsRuntimeDiskSpaceAvailable(runtime string) (bool, error) {
+	return dm.isSpaceAvailable(runtime, dm.policy.DockerFreeDiskMB, dm.cadvisor.ImagesFsInfo)
```
Comment (vishh) [inline pkg/kubelet/disk_manager.go]: Why not cache `runtime` in `realDiskSpaceManager` instead of passing it down?

File: pkg/kubelet/disk_manager.go (hunk 4 above)
Comment (vishh) [inline pkg/kubelet/disk_manager.go]: +1.

**Comment by vishh** [comment]
A couple of comments. Otherwise LGTM

File: pkg/kubelet/cadvisor/cadvisor_linux.go (hunk 5 above)
Comment (yifan-gu) [inline pkg/kubelet/cadvisor/cadvisor_linux.go]: @vishh So do we want to rely on cadvisor to provide image fs in the future info or not ?

File: pkg/kubelet/disk_manager.go (hunk 6 above)
Comment (sjpotter) [inline pkg/kubelet/disk_manager.go]: ok, doing that

File: pkg/kubelet/cadvisor/cadvisor_linux.go (hunk 5 above)
Comment (vishh) [inline pkg/kubelet/cadvisor/cadvisor_linux.go]: If cAdvisor source is working today, I'd leave it as-is.

**Comment by sjpotter** [comment]
just squashed this, but as it was all passed before, is it ready to go?

**Comment by sjpotter** [comment]
@k8s-bot test this github flake: #25919
